pints/_mcmc/_monomial_gamma_hmc.py

```diff
@@ -0,0 +1,507 @@
+    def _K(self, v_p, a, c, m):
+        """
+        Soft kinetic energy function defined in [1]
+        """
+        return np.sum([self._K_indiv(p, a, c, m) for p in v_p])
+
+    def _K_deriv_indiv(self, p, a, c, m):
+        """
+        Derivative of soft kinetic energy function defined in [1]
+        for individual momentum component
+        """
+        abs_p = np.abs(p)
+        sign_p = np.sign(p)
+        tanh = np.tanh(0.5 * c * abs_p**(1.0 / a) * sign_p / m)
+        return abs_p**(-2 + 1.0 / a) * p * sign_p * tanh / (a * m)
+
+    def _K_deriv(self, v_p, a, c, m):
+        """
+        Derivative of soft kinetic energy function defined in [1]
+        """
+        return np.array([self._K_deriv_indiv(p, a, c, m) for p in v_p])
+
+    def _pdf(self, p, a, c, m, z):
+        """
+        Auxillary kinetic energy probability density defined in [1]
+        """
+        return (1.0 / z) * np.exp(-self._K_indiv(p, a, c, m))
+
+    def _cdf(self, p, a, c, m, z):
+        """
+        Auxillary kinetic energy cumulative density defined in [1]
+        """
+        return integrate.quad(lambda p1: self._pdf(p1, a, c, m, z),
+                              -float('Inf'), p)[0]
+
+    def _inverse_cdf_calculator(self, a, c, m, z, pmax=100):
+        """
+        Estimates empirical cdf
+        """
+        p = np.linspace(-pmax, pmax, 1000)
+        lcdf = [self._cdf(p[i], a, c, m, z) for i in range(1000)]
+        f = interpolate.interp1d(lcdf, p, fill_value="extrapolate")
+        return f
+
+    def _sample_momentum(self):
+        """
+        Samples from soft kinetic energy pdf defined in [1]
+        """
+        us = np.random.uniform(size=self._n_parameters)
+        return np.array([self._f([u])[0] for u in us])
+
+    def set_epsilon(self, epsilon):
+        """
+        Sets epsilon for the leapfrog algorithm
+        """
+        epsilon = float(epsilon)
+        if epsilon <= 0:
+            raise ValueError('epsilon must be positive for leapfrog algorithm')
+        self._epsilon = epsilon
+        self._set_scaled_epsilon()
+
+    def epsilon(self):
+        """
+        Returns epsilon used in leapfrog algorithm
+        """
+        return self._epsilon
+
+    def scaled_epsilon(self):
+        """
+        Returns scaled epsilon used in leapfrog algorithm
+        """
+        return self._scaled_epsilon
+
+    def _set_scaled_epsilon(self):
+        """
+        Rescales epsilon along the dimensions of step_size
+        """
+        self._scaled_epsilon = np.zeros(self._n_parameters)
+        for i in range(self._n_parameters):
+            self._scaled_epsilon[i] = self._epsilon * self._step_size[i]
+
+    def ask(self):
+        """ See :meth:`SingleChainMCMC.ask()`. """
+        # Check ask/tell pattern
+        if self._ready_for_tell:
+            raise RuntimeError('Ask() called when expecting call to tell().')
+
+        # Initialise on first call
+        if not self._running:
+            self._running = True
```

Comment:
Could the code from initialise_KE not go here instead?

Reply:
Changed. Good shout!